src/internal/colors/HtmlColor.cpp

```diff
@@ -36,28 +36,35 @@ static inline char hexdigit(uint8_t v)
 }
 
 
+/**
+ * Generates a Html encoded hex color string (#aabbcc) with null termination.
+ *
+ * @param buf the buffer to write the string to
+ * @param bufSize the maximum buffer size (8 recommended for # + 123456 + null terminator)
+ * @return The amount of chars written to buf including the null terminator.
+ */
 size_t HtmlColor::ToNumericalString(char* buf, size_t bufSize) const
 {
-    size_t bufLen = bufSize - 1;
-
-    if (bufLen-- > 0)
+    if (bufSize > 0)
     {
-        if (bufLen > 0)
-        {
-            buf[0] = '#';
-        }
+        buf[0] = '#';
 
         uint32_t color = Color;
-        for (uint8_t indexDigit = 6; indexDigit > 0; indexDigit--)
+        size_t indexLast = min(7U, bufSize - 1);
+        if (indexLast < 7)
+        {
+            // note: the amount of digits may not match the buffer size and the most significant bits are on the left.
+            // For 0xaabbcc and a bufSize of 4 we want "#aa\0", not "#cc\0".
```

Comment:
The whole for() below that starts at last and walking backward was done to retain LSB when the buffer isn't large enough. That should be changed and not this extra stuff before it.

Reply:
I read your suggestion this way, that you want to reduce the operations inside the if as well and perform static calculations instead. The suggestion that I mean was this:

```cpp
    size_t indexLast = min(7, bufSize - 1);
    buf[indexLast] = '\0';
    for (size_t indexDigit = indexLast /* Starting at indexLast */; indexDigit > 0; --indexDigit) // note pre-decrement
    {
        buf[indexDigit] = hexdigit(color & 0x0000000f); /* Writing the least significant 4 bits to IndexDigit now, although it might not be complete */
        color >>= 4;
    }
```

Let's say color = 0xaabbcc and bufSize is 4, then we write c at indexLast, which is wrong, the result would be "#cc\0", but if someone wants a partial value it would rather be "#aa\0". One would probably want the first characters, as the "#" is being returned, too, no matter what buffer size > 1 they choose. If it would be the last characters it would rather be like "bcc\0" without "#", that's why MSB made more sense to me. That's what the original if inside the loop was for, but I reduced the steps in the loop instead and performed an initial shift of the color value to do that.

Yes, partial values will not happen very frequently and maybe we could also just return 0 then. I just wanted to make sure it is possible, if we can already specify the bufSize. Otherwise I don't see much sense in passing the bufSize at all, people could just make sure the buffer has a valid size outside. Or we could use `snprintf` and print the color with an ordinary C function.

Reply:
Passing the buf size is good practice at least for testing internally otherwise you get memory permission or worse corrupt memory issues.

Unfortunately, snprintf is not available on all Arduino platforms.  I understand the implementation of sprintf (the formatting part) is rather large for the small uC and would take up more flash space than some projects would want. 

I would vote for either keep it same (don't break existing uses) or change to not support partial at all.

Reply:
I've made another commit, you may want to squash them when merging (I think GitHub has the ability to do this). This time it strictly expects at least 8 characters and we don't need the lastIndex/indexLast variable anymore. Less RAM usage on Arduinos :-) .